src/coreclr/jit/rangecheck.cpp

```diff
@@ -1149,9 +1154,64 @@ void RangeCheck::MergeAssertion(BasicBlock* block, GenTree* op, Range* pRange DE
     else if (op->IsLocal())
     {
         assertions = block->bbAssertionIn;
+
+        // bbAssertionIn is a bit conservative and will not include intra-block assertions.
+        // e.g. created by GT_BOUNDS_CHECK nodes prior to the 'op' in the current block.
+        // For that, we walk the trees in the block until we find "op" and collect assertions
+        // along the way. It doesn't seem to be profitable for anything other than bounds checks
+        // (ArrBnds assertions), so we limit the search to only those blocks that may have bounds
+        // checks. It also helps reduce the cost of this search as it's not cheap.
+        //
+        // TODO-Review: EH successor/predecessor iteration seems broken.
```

Comment:
I don't see how this comment is relevant here. We are walking within a block from front to back. 

Looks like it comes from assertionprop and is (hopefully) quite stale?